man/push_puppetforge.8
.\" Manpage for push_puppetforge.
.\" Contact ggatward@redhat.com to correct errors or typos.
.TH PUSH_PUPPETFORGE 8 "04 Jan 2017" "sat6_scripts" "sat6_scripts User Manual" man page"
.SH NAME
push_puppetforge \- Export puppet forge modules and push to puppet-forge-server instance

.SH SYNOPSIS
.B push_puppetforge [\-o ORGANISATION] \-r REPO [\-s SERVER] [\-m MODULEPATH] [\-u USER]
.LP
.B "push_puppetforge --help"

.SH DESCRIPTION
.B push_puppetforge
is a script to push puppet forge modules that have been synchronised using Satellite 6
to an offline puppet forge server that uses the puppet-forge-server rubygem instance:
.br
.I https://github.com/unibet/puppet-forge-server
.LP
The puppet-forge-server instance requires a specific format for the module filesystem that
different from that used by Satellite 6. To make use of this script, the connected Satellite
must sync the puppet-forge repository, and have that repo included for the
.BR sat_export (8)
script. The disconnected Satellite must import the puppet modules into an identically named
puppet repository.
.LP
.BR push_puppetforge " will"
export the puppet modules from the disconnected Satellite and use rsync to transfer them to
the offline puppet-forge-server instance.

.SH OPTIONS
The options that apply to the
.B push_puppetforge
command are:
.PP
.BR "-o", " --org"
.I ORGANISATION
.RS 3
The organisation within Satellite to perform the import for. The default organisation is defined in the
.I config.yml
file, but can be overridden with this option.
.RE
.PP
.BR "-r", " --repo"
.I REPO
.RS 3
Label of the puppet-forge repository in Satellite to use as the sync source.
.RE
.PP
.BR "-s", " --server"
.I SERVER
.RS 3
Hostname of the puppet-forge-server instance.
.br
The default value is configured in the
.I config.yml
file
.RE
.PP
.BR "-m", " --modulepath"
.I MODULEPATH
.RS 3
Location of the puppet modules on the puppet-forge-server instance. (Default is /opt/puppet-forge/modules)
.RE
.PP
.BR "-u", " --user"
.I USER
.RS 3
Username to login to the puppet-forge-server as. Default is the user running the
.B push_puppetforge
script, from the
.I SUDO_USER
environment variable.
.RE
.PP
.BR "-p", " --password"
.I PASSWORD
.RS 3
API Token for user to post modules to Artifactory. Ignored if servertype is puppet-forge-server.
.RE

.SH FILES
.B Main Configuration
.RS 3
.I /usr/share/sat6_scripts/config/config.yml
.RE
.LP
.B Manifest download location
.RS 3
.I /var/sat-export/manifest/*
.RE

.SH SEE ALSO
.BR sat6_scripts_config (8),
.BR sat_export (8).

.SH AUTHOR
Geoff Gatward <ggatward@redhat.com>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Option to tag a published content view with a custom comment

### Fixed
- Fix unreferenced var in check_disk_space function
- Various minor README corrections


## [1.2.1] - 2017-12-11
### Fixed
- Incorrect mapping of days in auto_content.py
- Support for username:token authentication to Artifactory Puppet Forge server
- Email now supports multiple recipients


## [1.2.0] - 2017-12-10
### Added
- push_puppetforge now supports jFrog Artifiactory repository via HTTP POST
- sat_import now checks for exports that have not been imported (missed/skipped)
- sat_import --fixhistory option to force align import/export histories
- Email notification capability for use when automating content scripts
- Add unattended option to allow scripts to be automated
- auto_content scripts to allow unattended import/publish/promote/clean activity

### Changed
- --notar export saved in /cdn_export dir rather than /export to prevent it being deleted

### Removed
- Skip GPG short option (-n)


## [1.1.1] - 2017-10-25
### Added
- Allow limiting of Publish and Promotion to specified number of CV's at once.


## [1.1.0] - 2017-10-23
### Added
- Proxy support for download_manifest
- Quiet option in publish/promote scripts
- Full repo package comparison written to log during disconnected import
- Provide a history of successful content exports and imports
- Present a warning if import has already been performed for given dataset
- Option to leave export in unarchived format (--notar)
- Option to force a disconnected (import only) Satellite to export if required

### Changed
- Updated exit codes throughout the scripts
- Updated export dataset name format to allow for multiple exports per day
- Updated to Semantic Versioning

### Fixed
- Various minor bugfixes and typos
- Fixed bug in DoV export

### Deprecated
- Skip GPG short option (-n) will be removed in 1.2.0. Long option (--nogpg) will remain.


## [1.0] - 2017-03-06
- Initial 'public' release

## 0.6 - 2017-02-27
- Last of a series of pre-release betas

[Unreleased]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.2.1...HEAD
[1.2.1]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.2.0...1.2.1
[1.2.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.0...1.1.0
[1.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/0.6...1.0


push_puppetforge.py
#!/usr/bin/python
#title           :sat_export.py
#description     :Exports Satellite 6 Content for disconnected environments
#URL             :https://github.com/RedHatSatellite/sat6_disconnected_tools
#author          :Geoff Gatward <ggatward@redhat.com>
#notes           :This script is NOT SUPPORTED by Red Hat Global Support Services.
#license         :GPLv3
#==============================================================================
"""
Exports Satellite 6 yum content.
"""

import sys, argparse, datetime, os, shutil, pickle, re
import fnmatch, subprocess, tarfile
import simplejson as json
from glob import glob
import helpers

try:
    import yaml
except ImportError:
    print "Please install the PyYAML module."
    sys.exit(1)


def export_puppet(repo_id, repo_label, repo_relative, export_type):
    """
    Export Puppet modules
    Takes the type (full/incr)
    """
    numfiles = 0
    PUPEXPORTDIR = helpers.EXPORTDIR + '/puppet'
    if not os.path.exists(PUPEXPORTDIR):
        os.makedirs(PUPEXPORTDIR)

    msg = "Exporting Puppet repository id " + str(repo_id)
    helpers.log_msg(msg, 'INFO')

    # This will currently export ALL ISO, not just the selected repo
    msg = "Exporting all Puppet content"
    helpers.log_msg(msg, 'INFO')

    msg = "  Copying files for export..."
    colx = "{:<70}".format(msg)
    print colx[:70],
    helpers.log_msg(msg, 'INFO')
    # Force the status message to be shown to the user
    sys.stdout.flush()

    os.system('find -L /var/lib/pulp/published/puppet/http/repos/*' + repo_label \
        + ' -type f -exec cp --parents -Lrp {} ' + PUPEXPORTDIR + ' \;')

    # At this point the puppet/ export dir will contain individual repos - we need to 'normalise' them
    for dirpath, subdirs, files in os.walk(PUPEXPORTDIR):
        for tdir in subdirs:
            if repo_label in tdir:
                # This is where the exported ISOs for our repo are located
                INDIR = os.path.join(dirpath, tdir)
                # And this is where we want them to be moved to so we can export them in Satellite format
                # We need to knock off '<org_name>/Library/' from beginning of repo_relative and replace with export/
                exportpath = "/".join(repo_relative.strip("/").split('/')[2:])
                OUTDIR = helpers.EXPORTDIR + '/export/' + exportpath

                # Move the files into the final export tree
                if not os.path.exists(OUTDIR):
                    shutil.move(INDIR, OUTDIR)

                    os.chdir(OUTDIR)
                    numfiles = sum([len(files) for r, d, files in os.walk(OUTDIR)])
                    # Subtract the manifest from the number of files:
                    numfiles = numfiles - 1

                    # Since we are dealing with Puppet_Forge, create a second bundle for import to puppet-forge-server
                    if 'Puppet_Forge' in OUTDIR:
                        PFEXPORTDIR = helpers.EXPORTDIR + '/puppetforge'
                        if not os.path.exists(PFEXPORTDIR):
                            os.makedirs(PFEXPORTDIR)
                        os.system('find ' + OUTDIR + ' -name "*.gz" -exec cp {} ' + PFEXPORTDIR + ' \;')

                    msg = 'Puppet Export OK (' + str(numfiles) + ' files)'
                    helpers.log_msg(msg, 'INFO')
                    print helpers.GREEN + msg + helpers.ENDC

    # Now we are done with the original export dumps, we can delete them.
    shutil.rmtree(helpers.EXPORTDIR + '/export')
    shutil.rmtree(helpers.EXPORTDIR + '/puppet')

    return numfiles


def copy_to_pfserver(export_dir, pfserver, pfmodpath, pfuser):
    """
    Use rsync to copy the exported module tree to the puppet-forge-server instance
    """
    target = pfuser + '@' + pfserver + ':' + pfmodpath
    msg = 'Copying puppet modules to ' + target + '\n'
    helpers.log_msg(msg, 'INFO')
    print msg
    os.system('rsync -avrzc ' + export_dir + '/* ' + target)


def splitext(path):
    """
    Method to split the exported filename into author-module-version parts
    """
    for ext in ['.tar.gz']:
        if path.endswith(ext):
            return path[:-len(ext)], path[-len(ext):]
        return os.path.splitext(path)


def postModule(moduleTar, moduleInputDir, pfserver, pfmodpath, pfuser, pftoken):
    """
    Function to push puppet modules using curl to Artifiactory repository
    """
    # Remove module's extension (.tar.gz)
    puppetModuleNameNoExt = splitext(moduleTar)[0]

    # Remove the path from the module
    puppetModuleName = puppetModuleNameNoExt.split('/')[-1]

    # Split the module name into the required parts
    puppetModuleNameList = puppetModuleName.split('-')
    author = puppetModuleNameList[0]
    moduleName = puppetModuleNameList[1]
    version = puppetModuleNameList[2]

    url = "http://" + pfserver + pfmodpath + "/" + author + "/" + moduleName + "/" + moduleTar
    fileName = moduleInputDir + "/" + moduleTar

    # Put the files using curl (need to clean this up)
    authtoken = pfuser + ":" + pftoken
    subprocess.call(['curl', '-u', authtoken, '-XPUT', url, '-T', fileName])


def main(args):
    """
    Main Routine
    """
    #pylint: disable-msg=R0912,R0914,R0915

    # Who is running this script?
    runuser = helpers.who_is_running()

    # Set the base dir of the script and where the var data is
    global dir
    global vardir
    dir = os.path.dirname(__file__)
    vardir = os.path.join(dir, 'var')
    confdir = os.path.join(dir, 'config')

    # Check for sane input
    parser = argparse.ArgumentParser(description='Exports puppet modules in puppet-forge-server format.')
    # pylint: disable=bad-continuation
    parser.add_argument('-o', '--org', help='Organization (Uses default if not specified)',
        required=False)
    parser.add_argument('-r', '--repo', help='Puppetforge repo label', required=False)
    parser.add_argument('-t', '--type', help='Puppetforge server type (puppet-forge-server|artifiactory)', required=False)
    parser.add_argument('-s', '--server', help='puppet-forge-server hostname', required=False)
    parser.add_argument('-m', '--modulepath', help='path to puppet-forge-server modules',
        required=False)
    parser.add_argument('-u', '--user', help='Username to push modules to server as (default is user running script)',
        required=False)
    parser.add_argument('-p', '--password', help='Password (token) for username to push modules to Artifactory',
        required=False)
    args = parser.parse_args()

    # Set our script variables from the input args
    if args.org:
        org_name = args.org
    else:
       org_name = helpers.ORG_NAME

    # Define the type of puppet-forge server
    if args.type:
        pftype = args.type
    else:
        if not helpers.PFMETHOD:
            print "Puppet forge server type not specified"
            sys.exit(1)
        else:
            pftype = helpers.PFMETHOD

    # Define the puppet-forge-server hostname
    if args.server:
        pfserver = args.server
    else:
        if not helpers.PFSERVER:
            print "Puppet forge server not defined"
            sys.exit(1)
        else:
            pfserver = helpers.PFSERVER

    # Set the remote (puppet-forge-server) modules directory
    if args.modulepath:
        modpath = args.modulepath
    else:
        if not helpers.PFMODPATH:
            print "Puppet forge module path not defined"
            sys.exit(1)
        else:
            modpath = helpers.PFMODPATH

    # Set the username to use to push modules
    if args.user:
        pfuser = args.user
    else:
        pfuser = helpers.PFUSER

    # Read in the token for Artifiactory
    if args.password:
        pftoken = args.password
    else:
        pftoken = helpers.PFTOKEN

    # Record where we are running from
    script_dir = str(os.getcwd())

    # Get the org_id (Validates our connection to the API)
    org_id = helpers.get_org_id(org_name)

    # Read the repo label given by the user
    if args.repo:
        pfrepo = args.repo
    else:
        print "Puppetforge repo not defined"
        sys.exit(1)

    # Remove any previous exported content left behind by prior unclean exit
    if os.path.exists(helpers.EXPORTDIR + '/export'):
        msg = "Removing existing export directory"
        helpers.log_msg(msg, 'DEBUG')
        shutil.rmtree(helpers.EXPORTDIR + '/export')

    # Collect a list of enabled repositories. This is needed for:
    # 1. Matching specific repo exports, and
    # 2. Running import sync per repo on the disconnected side
    repolist = helpers.get_p_json(
        helpers.KATELLO_API + "/repositories/", \
                json.dumps(
                        {
                           "organization_id": org_id,
                           "per_page": '1000',
                        }
                ))


    # Process each repo
    for repo_result in repolist['results']:
        if repo_result['content_type'] == 'puppet':
            # If we have a match, do the export
            if repo_result['label'] == pfrepo:

                # Trigger export on the repo
                numfiles = export_puppet(repo_result['id'], repo_result['label'], repo_result['relative_path'], 'full')

            else:
                msg = "Skipping  " + repo_result['label']
                helpers.log_msg(msg, 'DEBUG')


    # Now we need to process the on-disk export data.
    # Define the location of our exported data.
    export_dir = helpers.EXPORTDIR + "/puppetforge"

    if (pftype == 'puppet-forge-server'):
        # Method for posting to puppet-forge-server
        os.chdir(script_dir)
        copy_to_pfserver(export_dir, pfserver, modpath, pfuser)

    elif (pftype == 'artifactory'):
        # Method for posting to Artifactory repository
        for module in os.listdir(export_dir):
            print("Posing: " + module)
            postModule(module, export_dir, pfserver, modpath, pfuser, pftoken)

    else:
        print("Unknown puppet-forge server type defined")
        sys.exit(1)

    # And we're done!
    print helpers.GREEN + "Puppet Forge export complete.\n" + helpers.ENDC
    sys.exit(0)

if __name__ == "__main__":
    try:
        main(sys.argv[1:])
    except KeyboardInterrupt, e:
        print >> sys.stderr, ("\n\nExiting on user cancel.")
        sys.exit(1)
